dbms/src/Interpreters/Expand2.cpp

```diff
@@ -0,0 +1,107 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Interpreters/Expand2.h>
+
+namespace DB
+{
+
+Expand2::Expand2(ExpressionActionsPtrVec projections_actions_, ExpressionActionsPtr before_expand_actions_, NamesWithAliasesVec projections_)
+    : leveled_projections_actions(projections_actions_)
+    , before_expand_actions(before_expand_actions_)
+    , leveled_alias_projections(projections_)
+{
+}
+
+Block Expand2::next(const Block & block_cache, size_t i_th_project)
+{
+    /// step1: clone a new block
+    ColumnsWithTypeAndName cloned_columns_and_type;
+    auto num_columns = block_cache.columns();
+    MutableColumns res(num_columns);
+    for (size_t i = 0; i < num_columns; ++i)
+    {
+        res[i] = block_cache.getColumns()[i]->cloneResized(block_cache.rows());
+    }
+    auto origin_column_and_types = block_cache.getColumnsWithTypeAndName();
+    for (size_t i = 0; i < num_columns; ++i)
+    {
+        cloned_columns_and_type.emplace_back(std::move(res[i]), origin_column_and_types[i].type, origin_column_and_types[i].name, origin_column_and_types[i].column_id, origin_column_and_types[i].default_value);
+    }
+    Block cloned_block(cloned_columns_and_type);
+
+    /// step2: execute a new block.
+    auto ith_projection = leveled_projections_actions[i_th_project];
+    ith_projection->execute(cloned_block);
+
+    /// step3: change raw name to unified alias name.
```

Comment:
Step3 actually picks some columns up and insert them into a new block, so I think this comment may confuse the reader.